Add example to test the smoothness/accuracy of the timestep

nfagerlund opened the pull request:
# Objective

Bevy's timestep is sometimes inaccurate, which can manifest as hiccups and stutters in what should
be smooth motion. This is most clearly visible when using pixel-perfect movement with large
visible pixels.

It would be good to have some kind of shared baseline for any conversations about hiccups, stutters, and other time-based jank!

## Solution

This PR adds a new example (in the stress tests directory) that scrolls a background behind a character at 120
pixels per second; by default, it makes sure to always move in whole-pixel increments. You can modify
some behavior while it's running by pressing keys:

- P: cycle between different `window::PresentMode`s. (NOTE: `Window.set_present_mode()` doesn't appear to actually work at the moment! Not sure what's up there and would appreciate any hints.) 
- W: cycle between different `window::WindowMode`s. (NOTE: `Window.set_resolution()` doesn't appear to do anything? I might be using it wrong?) 
- T: cycle between updating scroll position based on `Time.delta_seconds()`, and updating it by
  just assuming a 16.667 millisecond timestep (only useful in `PresentMode::Fifo` on a 60fps display).
- M: cycle between whole-pixel movement and simple sub-pixel movement.


---

## Changelog

Added a `time_smoothness` example (under `stress_tests`) for qualitatively evaluating the smoothness and accuracy of delta time. 

---

## Videos

<details><summary>A couple of videos of weird stuttering in action on Windows.</summary>

This computer is a desktop with a Haswell Core i3-4160 (entry-level processor from 2014), 8gb of ram, and a GTX 1060 (3gb) video card, running Windows 10 Home 21H2. It's quite modest! But it runs a wide range of commercial games with solid and respectable performance, and scrolling a background probably shouldn't look as bad as it does here. 

Windowed mode: 

https://user-images.githubusercontent.com/484309/170846616-08f63215-a017-4298-92a6-9da426cd11fc.mp4

Borderless fullscreen:

https://user-images.githubusercontent.com/484309/170846624-9c7c0f1e-1d7c-4022-bea4-989fe8e8f1d9.mp4

</details>

alice-i-cecile commented on examples/stress_tests/time_smoothness.rs:
```diff
@@ -0,0 +1,235 @@
+    window::{PresentMode, WindowMode},
+};
+
+// background is 905x363. To reduce empty space on each side and make sure
```
The value given here doesn't match that below. Can we remove the reference to the size of the background from this comment to avoid further desync?

alice-i-cecile reviewed (changes requested):
I'm impressed; this should be very useful! A few administrative notes:

1. Please add a module style doc comment to the top of your example.
2. Please add the example to examples/README.md

alice-i-cecile commented on examples/stress_tests/time_smoothness.rs:
```diff
@@ -0,0 +1,235 @@
+) {
+    let window = windows.primary_mut();
+    if input.just_pressed(KeyCode::P) {
+        // P: cycle PresentMode.
```
These controls should be mentioned at the start of the example, in the module style comment.

alice-i-cecile reviewed (changes requested):
Like I suggested on Discord, I'd like to see some numbers reported periodically in the console. To start, I think "mean, standard deviation, min and max frame time" for the last 10 seconds should be enough to give testers a better sense of what's going on.

You should be able to get this information from the FrameTimeDiagnosticsPlugin (I think in a resource); ping if you can't figure it out.

nfagerlund commented:
@alice-i-cecile Ok! I've added some frame time analysis to the example. I had to roll my own diagnostic, since it turns out the built-in one only keeps a non-configurable 20 frames, but now it prints this every 10 seconds: 

```
2022-06-08T06:21:55.395210Z  INFO time_smoothness: -------------------------
2022-06-08T06:21:55.395538Z  INFO time_smoothness: Average frame time: 0.016681
2022-06-08T06:21:55.395631Z  INFO time_smoothness: Standard deviation: 0.000569
2022-06-08T06:21:55.395689Z  INFO time_smoothness: Shortest frame:     0.006377
2022-06-08T06:21:55.395752Z  INFO time_smoothness: 95th percentile:    0.016797
2022-06-08T06:21:55.395840Z  INFO time_smoothness: 99th percentile:    0.017011
2022-06-08T06:21:55.395928Z  INFO time_smoothness: 99.5th percentile:  0.017293
2022-06-08T06:21:55.396014Z  INFO time_smoothness: Longest frame:      0.026963
2022-06-08T06:21:55.396077Z  INFO time_smoothness: -------------------------
```

I still need to go deal with Clippy, but figured I'd paste that before the output scrolls out of my terminal.

nfagerlund commented:
OK, all hail clippy, TIL about `.copied()`.

alice-i-cecile commented:
Awesome, that's super useful both as a snippet and as a testing tool.

hymm commented:
They were renamed/refactored in this PR https://github.com/bevyengine/bevy/pull/4745.

nfagerlund commented:
Oh, I also want to surface a question from the original post that maybe got lost — I have code in there that calls `Window.set_present_mode()` and `Window.set_resolution()`, and neither one of those seems to actually work. Anyone know what's up with those?

nfagerlund commented:
@hymm oh! Hmm, so the CI job must be merging or rebasing, instead of running on the branch. I'll rebase later and sort out the rename. thank you!

alice-i-cecile commented:
> Oh, I also want to surface a question from the original post that maybe got lost — I have code in there that calls Window.set_present_mode() and Window.set_resolution(), and neither one of those seems to actually work. Anyone know what's up with those?


IIRC you had to set this before calling .add_default_plugins, due to a quirk in the way `winit` owns the control loop :(

nfagerlund commented:
Oh... huh. I mean, I'm adding the window descriptor resource before that, so the initial settings take effect just fine. But so, can they just not be changed at runtime? Why would the functions on Window exist, then? I don't think you can even get hold of a Window struct until the app's running, yeah? 

Maybe I should remove those config bits and just tell users to edit and recompile. :/

aevyrie commented:
Could you change the stats to `ms` instead of `s`? Easier to read at those scales.

alice-i-cecile commented:
> But so, can they just not be changed at runtime? Why would the functions on Window exist, then? I don't think you can even get hold of a Window struct until the app's running, yeah?

This is about my feeling. Can you open an issue?

nfagerlund commented:
OK, I'm back!!

- I changed the frame time analysis to report in milliseconds. 
- I updated stuff that had fallen out of date with `main`, including the new camera types and the blurry-by-default texture smoothing. And I think I got the new examples metadata into shape correctly; we'll see in a sec.
- I filed an issue for those inert Window methods, https://github.com/bevyengine/bevy/issues/5111 -- it looks like they're _intended_ to work, so I left the example using them as advertised.

alice-i-cecile reviewed (changes requested):
Looks good! I'd be more comfortable merging this if you can drop a comment on the broken window commands linking to the issue in question.

I think it's *probably* worth leaving them in, if only to make it easy to test fixes to that. Other opinions would be welcome though.

Outside of that question, I'm very happy with this now.

hymm reviewed (commented):
![image](https://user-images.githubusercontent.com/2180432/176017525-db9f0a50-f78f-40e7-b66a-bde3672ba386.png)

The back of the Y in bevy is getting cut off for me on every other scroll by.

hymm commented on examples/stress_tests/time_smoothness.rs:
```diff
@@ -0,0 +1,355 @@
+const BG_WIDTH: f32 = 755.0;
+const BG_HEIGHT: f32 = 363.0;
+const BG_TILES: usize = 3;
+const BG_SPEED: f32 = 120.0;
```
Maybe add a recommendation here to change this number for different monitors. This is always going to hit the "big jump" warning on a 30Hz monitor and will probably always look jumpy on a 144Hz monitor.

richchurcher commented:
Closing as part of backlog cleanup: still seems like an interesting example, but would need significant revision and may not be as relevant as Bevy matures.
